lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -229,6 +229,61 @@ static Value createDecoder(Value input, unsigned width, Location insertLoc,
   return rewriter.create<DShlPrimOp>(insertLoc, resultType, bit, input);
 }
 
+/// Return the number of bits needed to index the given number of values.
+static size_t getNumIndexBits(ArrayRef<Value> values) {
+  return static_cast<size_t>(
+      std::max(std::ceil(std::log2(values.size())), 1.0) + 1);
+}
+
+/// Construct an arbiter based on a simple priority-encoding scheme. In addition
+/// to returning the arbiter result, the index for each input is added the the
+/// mapping for other lowerings to make use of.
+static Value createPriorityArbiter(ArrayRef<Value> inputs, Value defaultValue,
+                                   DenseMap<unsigned, Value> &indexMapping,
+                                   Location insertLoc,
+                                   ConversionPatternRewriter &rewriter) {
+  auto indexBits = getNumIndexBits(inputs);
+  auto indexType = SIntType::get(rewriter.getContext(), indexBits);
+  auto priorityArb = defaultValue;
+
+  for (unsigned i = inputs.size(); i > 0; --i) {
+    unsigned inputIndex = i - 1;
+
+    auto constIndex = createConstantOp(indexType, APInt(indexBits, inputIndex),
+                                       insertLoc, rewriter);
+
+    priorityArb = rewriter.create<MuxPrimOp>(
+        insertLoc, indexType, inputs[inputIndex], constIndex, priorityArb);
+
+    indexMapping[inputIndex] = constIndex;
+  }
+
+  return priorityArb;
+}
+
+/// Construct the logic to assign the ready outputs for ControlMergeOp and
+/// MergeOp. The logic is identical for each output. If the fired value is
+/// asserted, and the win value holds the output's index, that output is ready.
+static void createMergeArgReady(ArrayRef<Value> outputs, Value fired,
+                                Value winner,
+                                DenseMap<unsigned, Value> indexMappings,
+                                Location insertLoc,
+                                ConversionPatternRewriter &rewriter) {
+  auto bitType = fired.getType();
+
+  for (unsigned i = 0, e = outputs.size(); i != e; ++i) {
+    auto constIndex = indexMappings[i];
+
+    Value argReadyWire =
+        rewriter.create<EQPrimOp>(insertLoc, bitType, winner, constIndex);
+
+    argReadyWire =
+        rewriter.create<AndPrimOp>(insertLoc, bitType, argReadyWire, fired);
+
+    rewriter.create<ConnectOp>(insertLoc, outputs[i], argReadyWire);
+  }
+}
```

Comment:
I wasn't sure how much to refactor out of the lowerings for ControlMergeOp and MergeOp. I ended up pulling out the math and loops, but left all the straight-line code that generates wires and logic. It didn't seem worth it to have a helper for creating a wire named "resultDone" that was used twice.